### ISSUE TITLE
tracing: improve memory usage by using metadata-only child span types

### ISSUE THREAD

**Issue by irfansharif** [cockroachdb/cockroach#62020]
**Describe the problem**

#61777 was a late addition to 21.1 to bound the memory usage for the tracing package. It was necessary to address #59424, where we observed that creating real tracing spans for each statement resulted in a lot of memory pressure for not much real benefit. 

With TPC-C 1k, we were observing ~1000 root spans but around ~100k total spans. Most local root spans had ~270 children. Almost all of these child spans are completely empty, and really could be "metadata only". We can imagine a child span type that only held a pointer to the root span, and a span ID:

```go
type childSpan struct {
    spanID uint64 // and maybe operation name?
    rootSpan *Span
}
```

Today what we're doing is a lot more dire. Each child span is constructed using the full struct, which is quite large. It's comprised of the following structures:

https://github.com/cockroachdb/cockroach/blob/d7407d55814ddfb53d9d85a0e8cac0ba06d8da37/pkg/util/tracing/span.go#L49-L54

https://github.com/cockroachdb/cockroach/blob/d7407d55814ddfb53d9d85a0e8cac0ba06d8da37/pkg/util/tracing/span_inner.go#L26-L38

https://github.com/cockroachdb/cockroach/blob/4c76e19157ced57ba082ea0ab8cf5b317bdb5e1f/pkg/util/tracing/crdbspan.go#L29-L52

https://github.com/cockroachdb/cockroach/blob/4c76e19157ced57ba082ea0ab8cf5b317bdb5e1f/pkg/util/tracing/crdbspan.go#L58-L95

**Expected behavior**

That's all wildly excessive for most child spans that will never record anything, and could very well be "metadata only". Instead of https://github.com/cockroachdb/cockroach/issues/61407, but perhaps in the same spirit, we could "collapse" where the recordings are stored into the root span. In https://github.com/cockroachdb/cockroach/issues/59424#issuecomment-795237083, which an approximation of that idea, we observe that doing so can significantly relieve memory pressure (think "root span centered storage", which also has the nice property that there's no contention across traces). 

As for the pesky fact that in regular crdb code we often want to grab a handle on a span from the available context (`tracing.SpanFromContext`), we don't want the return type be an interface that abstracts over the root span and child span type. The reason for that is there's no way to define methods on nil interfaces, where as for structs we can:

```go
type Example struct {
}

func (e *Example) f() {
    if e == nil { }
}
```

Plumbing through interfaces would force callers to check whether or not the value is nil before doing anything. Luckily we have a work around. We could wrap the root span type in a stand-alone child span type that holds a reference to the root span, and has the same span ID. Given that, perhaps the type we want for the "child span" is `type Span`, and `type rootSpan` for the root, with `rootSpanInner` instead of today's `spanInner`.

**Comment by angelapwen** [comment]
I'm tying up loose ends as I offboard this week, and unassigning myself for now. Thanks again for filing this Irfan 😸

**Comment by irfansharif** [comment]
I tried some version of this as part of https://github.com/cockroachdb/cockroach/pull/64233 and realized I deluded myself writing this. While it's unfortunate child span objects are larger than they need to be, go's GC doesn't care too much about object size, just object count, and reducing total # of bytes allocated here doesn't much help performance. So it's unlikely we'll do anything here. What we did in  https://github.com/cockroachdb/cockroach/issues/59424#issuecomment-795237083 by collapsing all child spans into the root ends up foregoing all trace structure and really reduces the utility tracing provides.

> I'm not sure that I follow this.

That bit was more about not wanting to introduce a separate Span interface, working instead with the concrete types we have today. Reading it again, I'm not sure what I'm talking about either.
